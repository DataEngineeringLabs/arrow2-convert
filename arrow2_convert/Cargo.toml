[package]
name = "arrow2_convert"
version = "0.1.0"
authors = [
    "Jorge Leitao <jorgecarleitao@gmail.com>",
    "Chandra Penke <chandrapenke@gmail.com>"
]
edition = "2021"
license = "Apache-2.0 OR MIT"
keywords = ["Arrow", "arrow2"]
repository = "https://github.com/DataEngineeringLabs/arrow2-derive/arrow2_convert"
description = "Convert between nested rust types and Arrow with arrow2"

[dependencies]
arrow2 = { version = "0.9.1", default-features = false }
arrow2_convert_derive = { version = "0.1.0", path = "../arrow2_convert_derive", optional = true }
chrono = { version = "0.4", default_features = false, features = ["std"] }
err-derive = "0.3"
trybuild = "1.0"

[dev-dependencies]
arrow2_convert_derive = { version = "0.1.0", path = "../arrow2_convert_derive" }

[features]
default = [ "derive" ]
derive = [ "arrow2_convert_derive" ]
