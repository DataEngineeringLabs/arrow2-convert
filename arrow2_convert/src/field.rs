//! Implementation and traits for mapping rust types to Arrow types

use std::marker::PhantomData;

use arrow2::datatypes::{DataType, Field};
use chrono::{NaiveDate, NaiveDateTime};

/// Trait implemented by all types that can be used as an Arrow field.
///
/// Implementations are provided for types already supported by the arrow2 crate:
/// - numeric types: [`u8`], [`u16`], [`u32`], [`u64`], [`i8`], [`i16`], [`i32`], [`i128`], [`i64`], [`f32`], [`f64`],
/// - other types: [`bool`], [`String`]
/// - temporal types: [`chrono::NaiveDate`], [`chrono::NaiveDateTime`]
///
/// Custom implementations can be provided for other types.
///
/// The trait simply requires defining the [`ArrowField::data_type`]
///
/// Serialize and Deserialize functionality requires implementing the [`crate::ArrowSerialize`]
/// and the [`crate::ArrowDeserialize`] traits respectively.
pub trait ArrowField {
    /// This should be `Self` except when implementing large offset and fixed placeholder types.
    /// For the later, it should refer to the actual type. For example when the placeholder
    /// type is LargeString, this should be String.
    type Type;

    /// The [`DataType`]
    fn data_type() -> DataType;

    #[inline]
    #[doc(hidden)]
    /// For internal use and not meant to be reimplemented.
    /// returns the [`arrow2::datatypes::Field`] for this field
    fn field(name: &str) -> Field {
        Field::new(name, Self::data_type(), Self::is_nullable())
    }

    #[inline]
    #[doc(hidden)]
    /// For internal use and not meant to be reimplemented.
    /// Indicates that this field is nullable. This is reimplemented by the
    /// Option<T> blanket implementation.
    fn is_nullable() -> bool {
        false
    }
}

/// Enables the blanket implementations of [`Vec<T>`] as an Arrow field
/// if `T` is an Arrow field.
///
/// This tag is needed for [`Vec<u8>`] specialization, and can be obviated
/// once implementation specialization is available in rust.
#[macro_export]
macro_rules! arrow_enable_vec_for_type {
    ($t:ty) => {
        impl $crate::field::ArrowEnableVecForType for $t {}
    };
}
/// Marker used to allow [`Vec<T>`] to be used as a [`ArrowField`].
#[doc(hidden)]
pub trait ArrowEnableVecForType {}

// Macro to facilitate implementation for numeric types.
macro_rules! impl_numeric_type {
    ($physical_type:ty, $logical_type:ident) => {
        impl ArrowField for $physical_type {
            type Type = $physical_type;

            #[inline]
            fn data_type() -> arrow2::datatypes::DataType {
                arrow2::datatypes::DataType::$logical_type
            }
        }
    };
}

macro_rules! impl_numeric_type_full {
    ($physical_type:ty, $logical_type:ident) => {
        impl_numeric_type!($physical_type, $logical_type);
        arrow_enable_vec_for_type!($physical_type);
    };
}

// blanket implementation for optional fields
impl<T> ArrowField for Option<T>
where
    T: ArrowField,
{
    type Type = Option<<T as ArrowField>::Type>;

    #[inline]
    fn data_type() -> arrow2::datatypes::DataType {
        <T as ArrowField>::data_type()
    }

    #[inline]
    fn is_nullable() -> bool {
        true
    }
}

// u8 does not get the full implementation since Vec<u8> and [u8] are considered binary.
impl_numeric_type!(u8, UInt8);
impl_numeric_type_full!(u16, UInt16);
impl_numeric_type_full!(u32, UInt32);
impl_numeric_type_full!(u64, UInt64);
impl_numeric_type_full!(i8, Int8);
impl_numeric_type_full!(i16, Int16);
impl_numeric_type_full!(i32, Int32);
impl_numeric_type_full!(i64, Int64);
impl_numeric_type_full!(f32, Float32);
impl_numeric_type_full!(f64, Float64);

/// Maps a rust i128 to an Arrow Decimal where precision and scale are required.
pub struct I128<const PRECISION: usize, const SCALE: usize> {}

impl<const PRECISION: usize, const SCALE: usize> ArrowField for I128<PRECISION, SCALE> {
    type Type = i128;

    #[inline]
    fn data_type() -> arrow2::datatypes::DataType {
        arrow2::datatypes::DataType::Decimal(PRECISION, SCALE)
    }
}

impl ArrowField for String {
    type Type = Self;

    #[inline]
    fn data_type() -> arrow2::datatypes::DataType {
        arrow2::datatypes::DataType::Utf8
    }
}

/// Utf8 field that can be used by any type that can be converted from a String.
pub struct GenericUtf8<O: arrow2::types::Offset, S> {
    _data: PhantomData<S>,
    _o: PhantomData<O>,
}

impl<'a, S> ArrowField for GenericUtf8<i32, S>
where
    S: From<&'a str>,
{
    type Type = S;

    #[inline]
    fn data_type() -> arrow2::datatypes::DataType {
        arrow2::datatypes::DataType::Utf8
    }
}

impl<'a, S> ArrowField for GenericUtf8<i64, S>
where
    S: From<&'a str>,
{
    type Type = S;

    #[inline]
    fn data_type() -> arrow2::datatypes::DataType {
        arrow2::datatypes::DataType::LargeUtf8
    }
}

impl ArrowField for bool {
    type Type = Self;

    #[inline]
    fn data_type() -> arrow2::datatypes::DataType {
        arrow2::datatypes::DataType::Boolean
    }
}

impl ArrowField for NaiveDateTime {
    type Type = Self;

    #[inline]
    fn data_type() -> arrow2::datatypes::DataType {
        arrow2::datatypes::DataType::Timestamp(arrow2::datatypes::TimeUnit::Nanosecond, None)
    }
}

impl ArrowField for NaiveDate {
    type Type = Self;

    #[inline]
    fn data_type() -> arrow2::datatypes::DataType {
        arrow2::datatypes::DataType::Date32
    }
}

impl<'a> ArrowField for Vec<u8> {
    type Type = Self;

    #[inline]
    fn data_type() -> arrow2::datatypes::DataType {
        arrow2::datatypes::DataType::Binary
    }
}

/// Binary field that can be used by any type that can be converted from a [u8].
pub struct GenericBinary<O: arrow2::types::Offset, C> {
    _data: PhantomData<C>,
    _o: PhantomData<O>,
}

impl<'a, C> ArrowField for GenericBinary<i32, C>
where
    Self: 'a,
    C: FromIterator<u8>,
    &'a C: IntoIterator<Item = &'a u8>,
{
    type Type = C;

    #[inline]
    fn data_type() -> arrow2::datatypes::DataType {
        arrow2::datatypes::DataType::Binary
    }
}

impl<'a, C> ArrowField for GenericBinary<i64, C>
where
    Self: 'a,
    C: FromIterator<u8>,
    &'a C: IntoIterator<Item = &'a u8>,
{
    type Type = C;

    #[inline]
    fn data_type() -> arrow2::datatypes::DataType {
        arrow2::datatypes::DataType::LargeBinary
    }
}

/// FixedSizeBinary field that can be used by any type that can be converted from a [u8].
pub struct FixedSizeBinary<C, const SIZE: usize> {
    _data: PhantomData<C>,
}

impl<'a, C, const SIZE: usize> ArrowField for FixedSizeBinary<C, SIZE>
where
    Self: 'a,
    C: FromIterator<u8>,
    &'a C: IntoIterator<Item = &'a u8>,
{
    type Type = C;

    #[inline]
    fn data_type() -> arrow2::datatypes::DataType {
        arrow2::datatypes::DataType::FixedSizeBinary(SIZE)
    }
}

// Blanket implementation for Vec.
impl<T> ArrowField for Vec<T>
where
    T: ArrowField + ArrowEnableVecForType + 'static,
{
    type Type = Self;

    #[inline]
    fn data_type() -> arrow2::datatypes::DataType {
        arrow2::datatypes::DataType::List(Box::new(<T as ArrowField>::field("item")))
    }
}

/// List field that can be used by any type that can be converted from a [T].
pub struct GenericList<O: arrow2::types::Offset, C, T> {
    _d: std::marker::PhantomData<C>,
    _t: std::marker::PhantomData<T>,
    _o: std::marker::PhantomData<O>,
}

// Blanket implementation for Vec.
impl<'a, C, T> ArrowField for GenericList<i32, C, T>
where
    C: 'a,
    T: ArrowField + 'static,
    &'a C: IntoIterator<Item = &'a <T as ArrowField>::Type>,
    C: FromIterator<<T as ArrowField>::Type>,
{
    type Type = C;

    #[inline]
    fn data_type() -> arrow2::datatypes::DataType {
        arrow2::datatypes::DataType::List(Box::new(<T as ArrowField>::field("item")))
    }
}

impl<'a, C, T> ArrowField for GenericList<i64, C, T>
where
    C: 'a,
    T: ArrowField + 'static,
    &'a C: IntoIterator<Item = &'a <T as ArrowField>::Type>,
    C: FromIterator<<T as ArrowField>::Type>,
{
    type Type = C;

    #[inline]
    fn data_type() -> arrow2::datatypes::DataType {
        arrow2::datatypes::DataType::LargeList(Box::new(<T as ArrowField>::field("item")))
    }
}

/// FixedSizeList field that can be used by any type that can be converted from a [T].
pub struct FixedSizeList<C, T, const SIZE: usize> {
    _data: PhantomData<C>,
    _t: PhantomData<T>,
}

impl<'a, T, C, const SIZE: usize> ArrowField for FixedSizeList<C, T, SIZE>
where
    Self: 'a,
    T: ArrowField,
    &'a C: IntoIterator<Item = &'a <T as ArrowField>::Type>,
    C: FromIterator<<T as ArrowField>::Type>,
{
    type Type = C;

    #[inline]
    fn data_type() -> arrow2::datatypes::DataType {
        arrow2::datatypes::DataType::FixedSizeList(Box::new(<T as ArrowField>::field("item")), SIZE)
    }
}

arrow_enable_vec_for_type!(String);
arrow_enable_vec_for_type!(bool);
arrow_enable_vec_for_type!(NaiveDateTime);
arrow_enable_vec_for_type!(NaiveDate);

// Blanket implementation for Vec<Option<T>> if vectors are enabled for T
impl<T> ArrowEnableVecForType for Option<T> where T: ArrowField + ArrowEnableVecForType {}
