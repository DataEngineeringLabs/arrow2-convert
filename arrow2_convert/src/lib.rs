#![doc = include_str!("../README.md")]
#![deny(missing_docs)]
#![forbid(unsafe_code)]

pub mod deserialize;
pub mod field;
pub mod serialize;

// The proc macro is implemented in derive_internal, and re-exported by this
// crate. This is because a single crate can not define both a proc macro and a
// macro_rules macro.
#[cfg(feature = "arrow2_convert_derive")]
#[doc(hidden)]
pub use arrow2_convert_derive::ArrowField;

// Test README with doctests
#[doc = include_str!("../README.md")]
#[cfg(doctest)]
struct ReadmeDoctests;
