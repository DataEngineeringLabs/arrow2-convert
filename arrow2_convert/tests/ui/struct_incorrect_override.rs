use arrow2_convert::ArrowField;
use arrow2_convert::field::LargeBinary;

#[derive(Debug, ArrowField)]
struct Test {
    #[arrow_field(override="LargeBinary")]
    s: String
}

fn main() 
{}
