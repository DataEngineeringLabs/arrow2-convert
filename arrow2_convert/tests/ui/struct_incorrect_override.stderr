error[E0277]: the trait bound `String: Borrow<Vec<u8>>` is not satisfied
 --> tests/ui/struct_incorrect_override.rs:4:17
  |
4 | #[derive(Debug, ArrowField)]
  |                 ^^^^^^^^^^ the trait `Borrow<Vec<u8>>` is not implemented for `String`
  |
  = help: the following implementations were found:
            <String as Borrow<str>>
  = note: this error originates in the derive macro `ArrowField` (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0308]: mismatched types
 --> tests/ui/struct_incorrect_override.rs:4:17
  |
4 | #[derive(Debug, ArrowField)]
  |                 ^^^^^^^^^^ expected struct `String`, found struct `Vec`
  |
  = note: expected struct `String`
             found struct `Vec<u8>`
  = note: this error originates in the derive macro `ArrowField` (in Nightly builds, run with -Z macro-backtrace for more info)
